Make custom sync server preferences easier to understand

oakkitten opened the pull request:
This implements my own suggestion from the point 2 in Remaining issues in #12112 nearly verbatim, except I added a note about Sync url preference expecting the full URL as discussed in #12269. @dobefore What do you think?

<img src="https://user-images.githubusercontent.com/1710718/189485689-53f04b2a-de2f-4084-a0ec-ac947c8506b0.png" width=250> <img src="https://user-images.githubusercontent.com/1710718/189485686-aea5d070-9320-4e76-903e-15ef853d9d3a.png" width=250> <img src="https://user-images.githubusercontent.com/1710718/189485688-56ba32eb-f29d-475c-83a6-a7a780487abb.png" width=250> 

https://user-images.githubusercontent.com/1710718/189485733-83b66dcf-eb76-43d5-b062-9910e4b25b17.mp4

As usual, please see the commits for individual changes.

Note that in the first commit there's a link to the fragment https://docs.ankidroid.org/#_custom_sync_server which doesn't exist yet. I suppose the docs will have to be amended for the new release, and this section could be added later.

Also note that Accessibility scanner complains about two things; not sure if this is actionable.

  * Help preference summary: Link text. Consider using more descriptive text in the link. The link “Learn more” may not independently convey the link's purpose.
  * Sync url preference switch widget: Item descriptions. Multiple items have the same description. This clickable item's speakable text: "OFF" is identical to that of 1 other item(s).

Edit: The test [pass](https://github.com/oakkitten/Anki-Android/actions/runs/3028050757) in my repo.

## How Has This Been Tested?

Manually tested on my Xiaomi Mi A2 (Android 11) and Motorola Moto G 1st Gen (Android 6).
Added a unit test for preference migration.

## Checklist

- [ ] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [x] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [x] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

BrayanDSO reviewed (changes requested):
Overall I like it. It's a great improvement on this screen. Even being the least used screen, it's a nice progress for the app

BrayanDSO commented on AnkiDroid/src/main/res/values/10-preferences.xml:
```diff
@@ -206,12 +206,24 @@
+        You can specify either of the servers, or both;
+        if not set, the default AnkiWeb server will be used instead.
+        For example, you could use the sync server built into Anki
+        to synchronize the collection while you are off the grid,
```
I'm not a native, so I can't give a strong opinion, but is `off the grid` a common expression for stating that someone is offline?

oakkitten replied:
Also not a native. I used that to imply that someone is like not connected to the internet at large. “Offline” to me sounds more like not connected to _any_ networks, in which case syncing would be extremely difficult 😔

BrayanDSO replied:
I don't use custom sync server nor I have any experience with it, so WDYM by internet at large? Only connect to your local network, but not to the world wide web? If so, `off the grid` confuses me here, especially because further it says that someone can sync AnkiWeb to synchronize media online

oakkitten replied:
Yes, when you are off the grid and have no internet, you can still use synchronize with your Anki that's on your LAN. And then you can go to where internet is and sync your media using AnkiWeb. Kind of what Anki suggests

viciousAegis replied:
Off the grid is a common expression if you are a native speaker, but it can usually mean many different things like being not connected to the internet, being cut off from electricity, being lost in some remote part of the Amazon forest, or hiding from the secret service. I think a better way to phrase this would just be saying "while you are not connected to the internet"/ "your phone is not connected to the internet"

BrayanDSO replied:
"your phone is not connected to the internet" → remembering that tablets and TVs can run AnkiDroid as well

oakkitten replied:
> being lost in some remote part of the Amazon Forest, or hiding from the secret service

That's about what I had in mind when I was writing that 😛

mikehardy replied:
If you try to hide in the Amazon where I'm at, the Secret Service won't find you but the indigenous will ;-) https://www.sachalodge.com/indigenous-communities-of-the-ecuadorian-amazon/
Fun fact, whenever there is a general strike in Ecuador it's all a bunch of talk until the Amazon tribes join and walk to the capital then everyone gets real serious, real quick.
But you can learn their language using AnkiDroid, so you'll likely survive

oakkitten replied:
> But you can learn their language using AnkiDroid

Works best if you use the `amazonRelease` build!

viciousAegis replied:
> "your phone is not connected to the internet" → remembering that tablets, chromebooks and TVs can run AnkiDroid as well

We can then generalise using `Device` which should cover all bases

oakkitten replied:
As long as the phrase is common enough, I'd love to keep “off the grid”; it's short, it has the image of hiking in the woods that's less technical, and if you consider the rest of the sentence, I don't think you can misinterpret it.

> to synchronize the collection while you are off the grid, and AnkiWeb to synchronize media while online.

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/preferences/HtmlHelpPreference.kt:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    @Suppress("DEPRECATION") // Suggested alternative requires API 24
+    override fun getSummary() = super.getSummary()?.let { Html.fromHtml(it.toString()) }
```
I mostly followed the example of 56773418, with which I agree, since it's not deprecated because it's bad but deprecated because they added an argument. But I can change this if you want.

BrayanDSO replied:
I'd like so, yes. This way, a lint error is thrown if someone forgets to update this when minSdk 24 is reached

oakkitten replied:
Actually, there's `HtmlCompat` that... simply ignores flags on earlier APIs. Will do.

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/preferences/VersatileTextPreference.kt:
```diff
@@ -0,0 +1,85 @@
+        @Throws(Exception::class) fun validate(value: String)
+    }
+
+    val referenceEditText = AppCompatEditText(context, attrs)
```
I suppose not, but since the target is `AppCompatEditText` using one for reference leads to fewer questions I think?

BrayanDSO replied:
EditText is the default used on the app, and the default on androidx.Preference, so using a different type generates questions, yes, and as there isn't no need of using `AppCompatEditText`, should be better to just use `EditText`

oakkitten replied:
The actual view that androidx inflates is `AppCompatEditText`.

david-allison replied:
@BrayanDSO `com.ichi2.ui.FixedEditText` is the default for the app... because Android

BrayanDSO replied:
Here it's used only to get the inputType, not to serve on the UI, but nevermind

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/preferences/VersatileTextPreference.kt:
```diff
@@ -0,0 +1,85 @@
+                val positiveButton = alertDialog.getButton(DialogInterface.BUTTON_POSITIVE)
+
+                positiveButton?.isEnabled = try {
+                    validator.validate(editText.text.toString())
```
It is my unshakeable belief that verifiers should return an error. Do not ever create a verifier that returns a boolean. Or you risk ending up with 🐂💩 like `HostnameVerifier` and everyone who wants to get the actual error from a failing TLS collection is going to mention your name in sentences full of expletives.

Besides, what is used for verification thows, as it should, so you would have to have `try`/`catch` either way.

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/preferences/VersatileTextWithASwitchPreference.kt:
```diff
@@ -0,0 +1,100 @@
+class VersatileTextWithASwitchPreference(context: Context, attrs: AttributeSet?) :
+    VersatileTextPreference(context, attrs), DialogFragmentProvider {
+
+    init { widgetLayoutResource = R.layout.preference_widget_switch_with_separator }
```
Could add this but since we are an app and not a library, I'm not sure if this would be of much value. If need be, this can always be added later.

BrayanDSO replied:
Easy to do and more future proof, so it's very valid to me. But I wouldn't block on this, so it's implementer's choice.

oakkitten replied:
I would prefer KISS then—this can be super easily added later.

oakkitten commented on AnkiDroid/src/main/java/com/ichi2/anki/web/CustomSyncServer.kt:
```diff
@@ -20,39 +20,51 @@ import android.content.Context
 
-    fun getMediaSyncUrl(preferences: SharedPreferences): String? {
-        return preferences.getString(PREFERENCE_CUSTOM_MEDIA_SYNC_URL, null)
+    init {
+        AnkiDroidApp.getSharedPrefs(AnkiDroidApp.instance)
+            .registerOnSharedPreferenceChangeListener(preferenceChangeListener)
```
I would say that a listener that's registered and never unregistered is a more proper approach, or it would be if this wasn't a singleton. I am assuming that this entire class will get refactored into `Settings` or some equivalent of it when it is introduced, with proper dependency injection. But I can put it into `CustomSyncServerSettingsFragment` if you want

BrayanDSO replied:
> But I can put it into CustomSyncServerSettingsFragment if you want

Please do. Better to avoid any bugs

oakkitten replied:
Will do

oakkitten commented on AnkiDroid/src/main/res/values/10-preferences.xml:
```diff
@@ -206,6 +206,9 @@
     <string name="advanced_forecast_stats_mc_n_iterations_title" maxLength="41">Number of iterations of the simulation</string>
     <!-- Custom sync server settings -->
     <string name="custom_sync_server_title" maxLength="41">Custom sync server</string>
+    <string name="custom_sync_server_summary_none_of_the_two_servers_used">Not used</string>
```
Not sure what can I put in a comment that's not already in the identifier

BrayanDSO replied:
The same idea. It's easier for people to read a verbose and detailed comment then guessing just by a snake case key. assuming they read the key

oakkitten replied:
Will do

oakkitten commented on AnkiDroid/src/main/res/values/10-preferences.xml:
```diff
@@ -206,6 +206,9 @@
     <string name="custom_sync_server_title" maxLength="41">Custom sync server</string>
+    <string name="custom_sync_server_summary_none_of_the_two_servers_used">Not used</string>
+    <string name="custom_sync_server_summary_both_or_either_of_the_two_servers_used">Collection: %1$s\nMedia: %2$s</string>
+    <string name="custom_sync_server_summary_placeholder_default">default</string>
```
I thought about that and while I don't have a strong preference I like “default” more. If you read `Custom server? AnkiWeb!` it may sound a bit like maybe AnkiWeb is the name of a server application that you can install for yourself. “Default” also chimes a bit better with “Not used”, which I liked better than having “Default” if neither server is used. If you say “AnkiWeb” then “Not used” is a bit awkward... I guess in that case you'd also want to say “AnkiWeb” and that would be less ideal. Anyway, I'not a native so this is not a strong opinion.

BrayanDSO replied:
To me, it's more of a question that I have to think what `default` means, while `AnkiWeb` is straighthead, no thinking and no confusion. I'm not a native as well and I don't find having `Not used` first and `AnkiWeb` awkward at all. On my mind, their uses are distinct enough

oakkitten replied:
I think that in the context of custom x, it's rather clear that default x means not using the custom x. Or maybe not. Where are the natives when you need them? 😪

viciousAegis replied:
I think the confusion can be minimised if we just mention that AnkiWeb is the default server with something like:
`AnkiWeb (default)`

oakkitten replied:
I still dislike `AnkiWeb`, even with `(default)`, for the reasons mentioned above. I am willing to give in to peer pressure, but I request more peers (or one very assertive peer).

oakkitten commented on AnkiDroid/src/main/res/values/10-preferences.xml:
```diff
@@ -206,12 +206,24 @@
+        to synchronize the collection while you are off the grid,
+        and AnkiWeb to synchronize media while online.
+        <a href="https://docs.ankidroid.org/#_custom_sync_server">Learn more</a>
+        <br>
+        <br>Note that unlike in older AnkiDroid versions,
+        now you have to specify the full custom sync URL,
+        e.g. https://example.com:8080/sync/.]]></string>
```
I understand that this note is not exhaustive, but it is here to help people who would try putting `https://example.com:8080` here thinking it should work as it did before, as mentioned in #12269. I think it's better to keep this warning in some way.

As for the mandatory `/`, I hope that will get dealt with before the next release.

BrayanDSO replied:
Thought: could the url be checked to see if the provided url is setup correctly? Seems better from an UX perspective

oakkitten replied:
It's technically possible to sync with any valid URL, you don't necessarily have to have `/sync/` at the end.

BrayanDSO replied:
Exactly. My idea is to check if the connection stablished with the URL leads to an Anki server

oakkitten replied:
Not sure if this would be feasible given errors that AnkiDroid produces during syncing. Sounds complicated. You'd also need buttons that say “Test this URL” or something. Not sure if it's worth it.

BrayanDSO replied:
It would be an idea for another PR, which needs some polishing: Use a progress dialog like `withProgress` after OK is tapped. If successful, dismiss the dialog, else show a error message guiding how to fix it

You are one of the few people that actually use custom sync server, and the only one that cares enough to implement anything about it, so 100% up to if you want to do it or not.

oakkitten replied:
You might not be online when you are setting the URL... But yeah, in any case, this is best left for another PR.

BrayanDSO replied:
> You might not be online when you are setting the URL...

Another error message then

oakkitten replied:
It would be annoying and unexpected to have errors pop up when you simply want to set a setting. Maybe a dialog like this would work


    ┌─────────────────────────┐
    │  Sync url               │
    │                         │
    │  http://foo:132/sync/   │
    │                         │
    │  Test       Cancel  Ok  │
    └─────────────────────────┘

oakkitten commented on AnkiDroid/src/main/res/values/10-preferences.xml:
```diff
@@ -206,6 +206,19 @@
+        For example, you could use the sync server built into Anki
+        to synchronize the collection while you are off the grid,
+        and AnkiWeb to synchronize media while online.
+        <a href="https://docs.ankidroid.org/#_custom_sync_server">Learn more</a>
```
* Pinky promise to write it once the documentation is reshuffled to accommodate for the new preferences! 😛
* I'm also not sure about that. Does the documentation sever respect user language? I can add a comment but I'm not quite sure what to write there.

BrayanDSO replied:
I don't mind the timing of adding the docs, just want to be safe with the link. As I don't understand enough of ankidroiddocs, I can't help you much here. Probably better to wait for other reviewer, or post on discord.

Probably not very helpful: I know that a lot of links are extracted to constants, and IIRC something on the app used locales for something related to links

oakkitten replied:
I've looked at the strings and most are in resources with `tools:ignore="MissingTranslation"` so ¯\\\_(ツ)_/¯. I'll ask on Discord.

mikehardy replied:
ankiddroiddocs is not translated, there was some thinking around how to do that, but we went into a rabbit-hole of including translated manuals or not (tension is: it bloats the app, but Play store apparently has some APIs for that now, but then it's play store dependent... yuck) and status quo has just been to produce English docs and link the manual with "best effort" translations.

I'm fine with status-quo "add docs in English to ankidroiddocs and move on" since it's on the web at least and most people can auto-translate easily now

BrayanDSO replied:
I believe that the URL needs to be extracted from the string as a constant, so it can be changed in the future

oakkitten replied:
Hm, I wonder how I can do that. You can't reference a string from another string just like that [unless it's an XML entity or you are using a gradle plugin](https://stackoverflow.com/questions/4746058/reference-one-string-from-another-string-in-strings-xml) and I don't want to play with those. You can put a placeholder (`%s`) but then you would need to subclass the preference to actually insert the URL. I'm not sure if this is worth it?

It is unlikely that this URL is going to be reused in other strings, so maybe just rely on search-and-replace across files to change it if ever need be?

BrayanDSO replied:
AFAIK, currently we can't search and replace on all strings, because the translations are can't be uploaded from Crowdin (besides English)

oakkitten replied:
I should think Crowdin itself has a search and replace feature?

BrayanDSO replied:
I see that for individual files, but not an option for all languages, and manually replacing on 100 languages is quite boring. If there is a option to replace on all languages, I haven't found it.

Idk if the download-only will persist after AnkiDroid's tooling is updated to Crowdin new API, as I haven't reviewed that PR yet

oakkitten replied:
Well, I'm not sure what to do then.

(When I was using Crowdin I simply uploaded local translations to Crowdin. Also I think you can instruct Crowdin to not include untranslated strings when downloading from it.)

BrayanDSO replied:
The approach I like most is to keep it as is. Crowdin tooling should be improved in the future, and if this create an issue before the tooling is updated, it shouldn't break anything anyway. Someone may not even notice it as this screen is barely used

BrayanDSO commented on AnkiDroid/src/main/java/com/ichi2/anki/preferences/CustomSyncServerSettingsFragment.kt:
```diff
@@ -15,49 +15,50 @@
+
+    override fun onDestroy() {
+        super.onDestroy()
+        AnkiDroidApp.getSharedPrefs(AnkiDroidApp.instance)
```
As the fragment is a PreferenceFragmentCompat, you can get shared Preferences with `preference Manager.sharedPreferences` and drop the dependency on AnkiDroidApp, which eases testing, etc, etc

oakkitten replied:
Of course, that was silly of me. Done.

BrayanDSO commented on AnkiDroid/src/main/res/xml/preferences_custom_sync_server.xml:
```diff
@@ -6,19 +6,16 @@
-    <EditTextPreference
-        android:dependency="@string/custom_sync_server_enable_key"
+    <com.ichi2.anki.preferences.HtmlHelpPreference
+        android:summary="@string/custom_sync_server_help" />
```
Add `search:ignore="true"` to this preference, so it doesn't show on the search bar results.

`xmlns:search="http://schemas.android.com/apk/com.bytehamster.lib.preferencesearch"` should be necessary above

oakkitten replied:
Done

BrayanDSO commented:
Pending to me will be only
1.  the URL extraction to a constant from the string, so it may be changed in the future if necessary
2. Adding `search:ignore` to the HtmlHelpPreference

I'll leave the strings' texts to the natives.

oakkitten commented:
Rebased onto v2.16alpha85 to resolve a conflict and to add `search:ignore` to `HtmlHelpPreference`

mikehardy commented on AnkiDroid/src/main/res/values/10-preferences.xml:
```diff
@@ -206,12 +206,31 @@
+        is not used. E.g. “Collection: default” would mean that the default server,
+        that is AnkiWeb, will be used for synchronization.">default</string>
+    <string name="custom_sync_server_help"><![CDATA[
+        As an alternative to using AnkiWeb,
+        you can use your own servers for synchronization between devices.
+        You can specify either of the servers, or both;
+        if not set, the default AnkiWeb server will be used instead.
+        For example, you could use the sync server built into Anki
+        to synchronize the collection while you are off the grid,
+        and AnkiWeb to synchronize media while online.
+        <a href="https://docs.ankidroid.org/#_custom_sync_server">Learn more</a>
+        <br>
+        <br>Note that unlike in older AnkiDroid versions,
+        now you have to specify the full custom sync URL,
+        e.g. https://example.com:8080/sync/.]]></string>
```
```suggestion
        As an alternative to synchronizing with AnkiWeb,
        you may use your own servers for synchronization.
        You may specify either collection or media server, or both.
        If not set, the default AnkiWeb servers will be used instead.
        For example, you could use the sync server built into Anki
        Desktop to sync the collection when you only have a local network,
        and AnkiWeb to sync media when you have general internet access.
        <a href="%s">Learn more</a>
        <br>
        <br>Note: unlike in previous AnkiDroid versions,
        you must specify the full custom sync URL,
        for example: https://example.com:8080/sync/.]]></string>
```

**NOTE** I made the URL to our docs a format string parameter. We must be able to change those without perturbing translations

oakkitten replied:
The problem with `%s` is then you need an extra class that knows what to put in `%s`, or entity trickery, or something else like that.

mikehardy replied:
I understand it's difficult - but the URLs move. They *will* move - https://github.com/ankidroid/ankidroiddocs/pull/103 - URLs must not be in translated strings unfortuunately

oakkitten replied:
Found this https://github.com/LikeTheSalad/android-stem. Worth it? If not, I'm thinking of making the preference take a few arguments such as `app:substitution_1`

BrayanDSO replied:
Can't you just generate the string and set it inside `initSubscreen` on the settings fragment? It's similar to what is done with the "Anki card" and "Card browser" context menu preferences

Although I like the plugin you linked. How much does it affect build time?

oakkitten replied:
I'd prefer a more generic solution, it's not hard to do. I did not test that plugin. It looks pretty good, but I wouldn't want to add it just for one string. I suggest considering using it if there are more strings that can be inlined in a similar fashion.

BrayanDSO replied:
From head, there are the URLs on the About screen and the "Anki card" and "Card browser" strings.

I bet that there are some other strings with URLs that could be benefitted from this as well

oakkitten replied:
I tried using android-stem and it works but reuses old output unless build caches are cleared. I'm not good enough with Gradle to investigate. Might be our fault? Filed an issue here https://github.com/LikeTheSalad/android-stem/issues/35

Otherwise, plugin's tasks take about 7-10 seconds on my SLOW machine if the strings are changed. If no strings are changed it's a no op.

mikehardy replied:
I'm okay with a plugin - the issue you raised will be a good test of how it's maintained. The only problem is if this becomes a hard build requirement and then with gradle 9 (or whatever) it stops being maintained. Now that I think of it, that might be inevitable. If your minimum viable sketch for an implementation that allowed the substitution in our code was not too hard to do, might be better.

Sorry for the constraint, I just know that URLs will move, it's inevitable

oakkitten replied:
It's a one-man project so the danger is there alright. But other than this one issue it seems to be really neat. The approach is really intuitive. I checked the About screen at it can help there; and if this kind of `HtmlHelpPreference` is used elsewhere in AnkiDroid it would be useful in other “Learn more” links.

OTOH implementing `app:substitution_1` is not hard and will cover the current and future uses of `HtmlHelpPreference`, and the About screen already works, and is not bad as it is.

Dunno which way is better. ¯\\\_(ツ)_/¯

mikehardy replied:
Let's go with the no-dependency route if it's a toss-up, to bring in a dependency that is nice-to-have but not hard to live without doesn't have enough value I think, though I do respect one-person efforts I know personally (as the creator of many one-person efforts...) that the strain is  real and I would not be surprised at all if the person was like "nah, I'm done here", whereas we are setting up for our second decade...

oakkitten replied:
Aight, done in 99943f6

mikehardy replied:
I appreciate it, thank you

mikehardy reviewed (changes requested):
I like all of this except:

- I liked the language but tried to make it even more precise + prescriptive
- we must maintain URL flexibility without trashing translations, so that should be interpolated

mikehardy commented:
And took a conflict during my current merge-fest, apologies. Probably an AlertDialog thing :see_no_evil:

oakkitten commented:
Rebased on current main to resolve conflicts, also incorporated working changes into the first commit and added a commit extracting the URL into a separate string. Now if the tests would be so kind to pass...

BrayanDSO reviewed (changes requested):
The last changes LGTM. 

Noticed only now, after reading `HtmlHelpPreference` javadoc, that the new preferences on `com.ichi2.anki.preferences`, but all the other individual `Preference` classes are on `com.ichi2.preferences`, so they should be moved.

oakkitten commented:
Good catch! Rebased. (Rebasing with renaming is hell!) 

I put the helper methods into a new file `com/ichi2/preferences/PreferenceUtil.kt` since they belong with the preferences. They were in `com/ichi2/preferences/anki/PreferenceUtils.kt` before. Didn't have a better name for the file, opted to change Utils to Util so it's not the same.

BrayanDSO reviewed (approved):
*L*ooks
*G*ood
*T*o
*M*e

Actually not, lint error

mikehardy reviewed (approved):
I believe we'll have to squash merge just on the off-chance that there is an attempt to do a non-whole-PR-atomic revert as the original string then the final string-that-has-URL-substituted cannot be separated or crowdin would be in a weird state, but I'm okay with that.

Please Please Please if someone else merges this follow the string instructions comment and do an I18N sync+merge *first*, *then merge this* then immediately *do another i18n sync*. Well "immediately" is "waiting 45mins between syncs to make sure crowdin doesn't throttle the new string bundle build, but that's usually not a big limiter.

The activity stream on crowdin is also a good indicator of the inbound translation count

oakkitten commented:
Not at computer right now but I can maybe merge the last commit into the first? I hate those squash commits vehemently

mikehardy commented:
I hear ya - your commits do tell a story, there's no hurry on this, I'll remove the pending merge label, take your time and re-order as desired. Scoped storage still obviously the priority I just had a bit of extra time so looked at all PRs vs just scoped storage...

mikehardy commented:
and a conflict, just for extra fun :see_no_evil:

oakkitten commented:
Done, and rebased on current main to resolve conflict.

mikehardy commented:
Great - running through `./tools/test-commits.sh <hash>` on the commit stream now, and will do a strings sync to clear things out for all the changes here

mikehardy reviewed (changes requested):
Well...that's weird, on the very first commit `./gradlew ktlintFormat` fails with:

```
/home/mike/work/AnkiDroid/Anki-Android-Upstream/AnkiDroid/src/main/java/com/ichi2/preferences/PreferenceUtil.kt
interface DialogFragmentProvider should be declared in a file named DialogFragmentProvider.kt 
```

...but....it works on the other 5 commits (!?)

Something about the extra sauce added in the following commit: https://github.com/ankidroid/Anki-Android/pull/12367/commits/d0e1209071ac2b66a8e4bd5a662713b41f06d2e1 makes ktlint okay with the interface in file not named after the interface :thinking: 

Could either dig deep and learn a bunch of things that may never be needed again to fix it, or just squash those two commits and be done, or squash them all :laughing: 

Let me know what you think

oakkitten commented:
I assume this rule is supposed to catch typos, like when you have a class `FooBar` in `FooBaz.kt` or something. Rebased to reorder commits 1 & 2 that should remove the lint error. 

(Why squash though? I'm not exactly sure why one of commits failing lint is an issue. As long as you don't merge with `--ff-only`—I prefer `--no-ff` for almost every merge—you get a merge commit and you can revert a merge commit. You can eat a cake and have it too eh?

I'd even go so far to say there's not much value in ensuring every commit builds. If something breaks you can still use `git blame` with `--first-parent`, and if this finds that the offending commit comes from a feature branch that has broken commits, well, fixing those commits while doing `git blame` is about the same amount of work as doing it right away. Of course, if you happen to find that the feature is broken in another `git blame` you'll have to redo the work but what are the chances of that.)

mikehardy commented:
- No idea what ktlint is unhappy with at a technical level here
- github doesn't allow FF control when using web UI and we're using the web UI so it's a constraint we're stuck with
- we have tools that git bisect, every commit building is a useful thing

will re-check...

oakkitten commented:
* I think that rule is not like an actual lint rule but more of a warning
* It doesn't allow `--ff-only`, but I think the default merge strategy is `--no-ff`, so as long as you don't squash or rebase it will make a neat merge commit
* I was actually talking about `git bisect` (I always confuse the two!), it also has `--first-parent`

mikehardy reviewed (approved):
`/bin/nice -n 19 ./tools/test-commits.sh 0746ab0b4b08a74b6b7f609703d2628c496e8c35` clean

mikehardy commented:
merge commit is not currently enabled for the repo, I've got squash or rebase - this is "rebase clean" now, so I'm going with that - will do the strings sync after
